[mlir][transform] Overhaul `RegionBranchOpInterface` implementations.

ingomueller-net opened the pull request:
This PR adds the `ReturnLike` trait to `transform.yield`. This is required in the one-shot bufferization pass since the merging of  #110322, which analyses any `FunctionOpInterface` and expects them to have a `ReturnLike` terminator.

If we give `transform.yield` the `ReturnLike` trait, then the checks made by `RegionBranchOpInterface` behave differently. More concretely, the check that the operands and results passed across the control flow edges from the parent op to its regions and back are equal (or compatible). Which results are passed back from regions to the parent op depend on the terminator and/or whether it is `ReturnLike`.

This commit radically changes the implementations of the `alternatives`, `foreach`, and `sequence` ops, which are the ops that use `yield` as their terminator. In fact, all of these ops only ever pass control from the parent op to the region and from there back to the parent op. In particular and unlike `scf.for`, `transform.foreach` does *not* pass control from one iteration of its body to the next directly; it rather passes the control back to the parent op, which then passes it back to the body for the next iteration. That can be seen by the fact that the body always gets arguments of the same type as the operands of the parent op (and none of the yielded types) and the types that are yielded correspond exactly to the result types of the parent op.

It is unclear why the previous implementation worked. Clearly, the type checks have not been executed because the terminator was not `ReturnLike`, so the mismatching types passed unnoticed. I suppose that there simply is/was no other use case of using the `RegionBranchOpInterface` of the affected ops yet.

ingomueller-net committed:
[mlir][transform] Make `yield` a `ReturnLike` op.

This PR adds the `ReturnLike` trait to `transform.yield`. This is
required in the one-shot bufferization pass since the merging of
 #110332, which analyses any `FunctionOpInterface` and expects them to
have a `ReturnLike` terminator.

Signed-off-by: Ingo Müller <ingomueller@google.com>

ingomueller-net commented:
One could also argue that the one-shot bufferization is using the interface wrong: AFAIU, `FunctionOpInterface` does not prescribe the ops implementing it to have `ReturnLike` terminators, so the pass should not assume this kind of terminators.

matthias-springer reviewed (approved):
Irrespective of the bufferization issue, this change makes sense to me. `yield` ops in other dialects typically implement the `ReturnLike` trait, e.g., `scf.yield`.

ingomueller-net committed:
Overhaul `RegionBranchOpInterface` impls of ops using `yield`.

If we give `transform.yield` the `ReturnLike` trait, then the checks
made by `RegionBranchOpInterface` behave differently. More concretely,
the check that the operands and results passed across the control flow
edges from the parent op to its regions and back are equal (or
compatible). Which results are passed back from regions to the parent op
depend on the terminator and/or whether it is `ReturnLike`.

This commit radically changes the implementations of the `alternatives`,
`foreach`, and `sequence` ops, which are the ops that use `yield` as
their terminator. In fact, all of these ops only ever pass control from
the parent op to the region and from there back to the parent op. In
particular and unlike `scf.for`, `transform.foreach` does *not* pass
control from one iteration of its body to the next directly; it rather
passes the control back to the parent op, which then passes it back to
the body for the next iteration. That can be seen by the fact that the
body always gets arguments of the same type as the operands of the
parent op (and none of the yielded types) and the types that are yielded
correspond exactly to the result types of the parent op.

It is unclear why the previous implementation worked. Clearly, the type
checks have not been executed because the terminator was not
`ReturnLike`, so the mismatching types passed unnoticed. I suppose that
there simply is/was no other use case of using the
`RegionBranchOpInterface` of the affected ops yet.

Signed-off-by: Ingo Müller <ingomueller@google.com>

ingomueller-net commented:
I just pushed a new commit (hopefully) fixing the failed tests. That commit increases the magnitude of the commit significantly, so I'd like another full review before proceeding.

The new commit changes the behavior of the implementations of `RegionBranchOpInterface` quite radically, to the point where I wonder whether the previous implementations were at all functional. I couldn't reconstruct what they were trying to achieve but they did and do not fit my mental model of what the ops do. See the commit message for details.

matthias-springer commented on mlir/lib/Dialect/Transform/IR/TransformOps.cpp:
```diff
@@ -104,16 +104,8 @@ transform::AlternativesOp::getEntrySuccessorOperands(RegionBranchPoint point) {
 
 void transform::AlternativesOp::getSuccessorRegions(
     RegionBranchPoint point, SmallVectorImpl<RegionSuccessor> &regions) {
-  for (Region &alternative : llvm::drop_begin(
```
Removing this does not seem right. When a region fails, the op jumps to the next region.

matthias-springer replied:
`In particular and unlike scf.for, transform.foreach does not pass control from one iteration of its body to the next directly; it rather passes the control back to the parent op, which then passes it back to the body for the next iteration. That can be seen by the fact that the body always gets arguments of the same type as the operands of the parent op (and none of the yielded types) and the types that are yielded correspond exactly to the result types of the parent op.`

I don't follow this part. Where is the type mismatch here?

matthias-springer replied:
Oh, I guess you are saying that the successor operands do not match with the block arguments of the next region.

matthias-springer replied:
I think the `RegionBranchOpInterface` does not support the use case here. When you branch back to the parent op, there is no way to go back into the op a second time.

```
    A "region successor" indicates the target of a branch. It can indicate
    either a region of this op or this op. In the former case, the region
    successor is a region pointer and a range of block arguments to which the
    "successor operands" are forwarded to. In the latter case, the control flow
    leaves this op and the region successor is a range of results of this op to
    which the successor operands are forwarded to.
```

This sentence talks about results of the op, so if we would allow going back into the op, you could set the same results multiple times.
```
    In the latter case, the control flow
    leaves this op and the region successor is a range of results of this op to
    which the successor operands are forwarded to.
```

I think we should clarify this in the documentation of `RegionBranchOpInterface`. (We could also allow it but most transformations on top of `RegionBranchOpInterface` probably assume that you cannot go back into the op.)

Do we have to implement the `RegionBranchOpInterface` on these transform ops at all?

ingomueller-net replied:
In `sequence`, I think the sentences matches, no? When control reaches `yield`, the operands of yield are forwarded to the parent op, they become the result of the op, and "control flow leaves [the] op," right?

What doesn't quite fit for `foreach` and `alternatives` is the "control flow leaves [the] op" part. One could argue that the "region successor is a [part of or a potential] range of results of [the] op," but then the control flow does *not* leave the op; instead it may go to the next region or re-enter the body. Plus there are rules about how to combine the different results yielded by different regions/iterations (though the types always match).

ingomueller-net replied:
> Do we have to implement the `RegionBranchOpInterface` on these transform ops at all?

Maybe not. I just removed it from the three ops and the tests in upstream still pass.

matthias-springer replied:
> Plus there are rules about how to combine the different results yielded by different regions/iterations (though the types always match).

I think that alone is already reason enough to drop the `RegionBranchOpInterface` from `scf.forall`.

ingomueller-net replied:
> > Plus there are rules about how to combine the different results yielded by different regions/iterations (though the types always match).
> 
> I think that alone is already reason enough to drop the `RegionBranchOpInterface` from `scf.forall`.

...from `transform.foreach`, right?

ingomueller-net replied:
And from `transform.alternatives` as well, no?

matthias-springer replied:
Yes.

I just took a look at `scf.forall`:

```
void ForallOp::getSuccessorRegions(RegionBranchPoint point,
                                   SmallVectorImpl<RegionSuccessor> &regions) {
  // Both the operation itself and the region may be branching into the body or
  // back into the operation itself. It is possible for loop not to enter the
  // body.
  regions.push_back(RegionSuccessor(&getRegion()));
  regions.push_back(RegionSuccessor());
}
```

Maybe the entire modeling of operands could be optional.

ingomueller-net replied:
I am not sure I understand what you mean. AFAIU, `RegionSuccessor()` creates a `RegionSuccessor` that `isParent`, i.e., the function indicates that the region may give control flow back to the parent op. The fact that there are no arguments provided means that it gives the control back with passing results on the way. That's different from not saying which arguments it passes, no?

matthias-springer replied:
I was wondering about this line:
```
regions.push_back(RegionSuccessor(&getRegion()));
```
A region is specified but no basic block arguments.

So I am trying to understand if this interface can be used to model branches between regions, without modeling any data flow via block arguments. (I think it is possible.)

ingomueller-net replied:
For that line, I think the same applies as what I said about the other line: this constructs the `RegionSuccessor` with an empty `BlockArgListType` via a default argument of [this constructor](https://github.com/llvm/llvm-project/blob/d626f95564/mlir/include/mlir/Interfaces/ControlFlowInterfaces.h#L169). And I believe that that means "no arguments are passed" rather than "it isn't specified what arguments are passed."

ingomueller-net replied:
I believe that "it isn't specified which arguments are passed" cannot be expressed currently. An empty argument list is always interpreted as "no arguments are passed."

matthias-springer replied:
Yes, that's correct. But I am wondering if it makes any difference for analyses and transformations.

If we a region has an argument, but the `RegionBranchOpInterface` says that no arguments are forwarded to that region: what does that mean? It means that we have no idea where the data for that block argument is coming from. Maybe the op itself produces it.

Or maybe the value was actually forwarded from another region but we did not account for it in the `RegionBranchOpInterface` (and the terminator interface).

Does it matter which one is the case for an analysis that checks the `RegionBranchOpInterface`? It has to be conservative around such cases anyway.

ingomueller-net replied:
> If we a region has an argument, but the `RegionBranchOpInterface` says that no arguments are forwarded to that region: what does that mean?

Are you sure that that can even exist? I think I ran into failed type check that seem to check exactly that while working on the last commit...

matthias-springer replied:
I think scf.forall is an example.

ftynse reviewed (changes requested):
Sorry, I missed this. This may depend on what we understand by `ReturnLike`, but `transform.yield` is likely not that. In particular, `ReturnLike` implies one-to-one mapping of the yielded value to the result of the parent operation or the arguments of the next region in dataflow analyses, which is not the case for `transform.yield` used in `alternatives`, or `foreach`.

ingomueller-net commented:
> One could also argue that the one-shot bufferization is using the interface wrong: AFAIU, `FunctionOpInterface` does not prescribe the ops implementing it to have `ReturnLike` terminators, so the pass should not assume this kind of terminators.

Thanks, @ftynse, for helping us decide. I think that my earlier statement (quoted here) is then correct: one cannot assume that `FunctionOpInterface` ops have `ReturnLike` terminators, so that change was not valid. That also corresponds to @ftynse's comment in #112615.
